Rustdoc stable urls

jyn514 opened the pull request.

jyn514 committed:
stable rustdoc urls

jyn514 committed:
More links!

Nemo157 commented on text/0000-stable-rustdoc-urls.md:
```diff
@@ -0,0 +1,279 @@
+[motivation]: #motivation
+
+[Rustdoc] generates a separate HTML page for each [item] in a crate.
+The URL for this page currently stable relative to rustdoc; in other words,
```
```suggestion
The URL for this page is currently stable relative to rustdoc; in other words,
```

jyn514 committed:
Fix todo

Nemo157 commented on text/0000-stable-rustdoc-urls.md:
```diff
@@ -0,0 +1,280 @@
+since function locals cannot be documented.
+
+1. The value namespace. This includes `fn`, `const`, and `static`.
+2. The type namespace. This includes `mod`, `struct`, `union`, `enum`, `trait`, `type`, struct fields, and enum variants.
```
```suggestion
2. The type namespace. This includes `struct`, `union`, `enum`, `trait`, and `type`.
```

* struct fields are not at the module level
* `doc(inline)` on enum variant re-exports does not work
* `mod`s have folders rather than files (this one I'm 50/50 on including in the list)

jyn514 replied:
Ah hmm, maybe I should move this to the section on URL fragments instead. I want to keep `mod` - I'll mention that it uses a folder instead of getting its own page.

jyn514 committed:
grammar and typos

Nemo157 commented on text/0000-stable-rustdoc-urls.md:
```diff
@@ -0,0 +1,280 @@
+3. The macro namespace. This includes `macro_rules!`, attribute macros, and derive macros.
+
+Rust does not permit there to be overlaps within a namespace,
+except for overlaps caused by globbing, where there must be a clear 'primary' import.
```
There are no overlaps, globbed imports are shadowed and blocked by non-glob imports and so don't actually exist (e.g. in <https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=548f2f5e08600d4ad732c407ab3dd59f> the trait is not imported).

jyn514 committed:
Fill out 'alternatives' section

jyn514 committed:
Add 'prior art' and 'unresolved questions'

jyn514 committed:
Address review comments

- Address `mod` specially
- Move enum variants and structs to the 'URL fragments' section
- Fix misleading wording around glob imports

jyn514 committed:
typo

Co-authored-by: Nemo157 <github@nemo157.com>

GuillaumeGomez commented on text/0000-stable-rustdoc-urls.md:
```diff
@@ -0,0 +1,317 @@
+
+Rustdoc will use the following links, depending on the namespace:
+
+- `Name.t.html` for types
```
I'd suggest instead: `t.Name.html`.

GuillaumeGomez replied:
The kind being first is what I meant. ;)

jyn514 committed:
name.namespace -> namespace.name

jyn514 committed:
Remove commented out text

jyn514 commented:
Closing, I opened a pre-rfc: https://internals.rust-lang.org/t/pre-rfc-stable-rustdoc-urls/13099/1
